packages/frontend/editor-ui/src/components/RunDataAi/utils.test.ts

```diff
@@ -189,6 +189,254 @@ describe(getTreeNodeData, () => {
+		expect(rootNode1Tree[0].children.length).toBe(1);
+		expect(rootNode1Tree[0].children[0].node).toBe('SharedSubNode');
+		expect(rootNode1Tree[0].children[0].runIndex).toBe(0);
+
+		// Test for RootNode2 - should only show SharedSubNode with source RootNode2
+		const rootNode2Tree = getTreeNodeData('RootNode2', workflow, aiData);
+		expect(rootNode2Tree[0].children.length).toBe(1);
+		expect(rootNode2Tree[0].children[0].node).toBe('SharedSubNode');
+		expect(rootNode2Tree[0].children[0].runIndex).toBe(1);
+	});
+
+	it('should include nodes without source information', () => {
+		const workflow = createTestWorkflowObject({
+			nodes: [createTestNode({ name: 'RootNode' }), createTestNode({ name: 'SubNode' })],
+			connections: {
+				SubNode: {
+					ai_tool: [[{ node: 'RootNode', type: NodeConnectionTypes.AiTool, index: 0 }]],
+				},
+			},
+		});
+
+		// Create test AI data with a node that has no source information
+		const subNodeData = {
+			node: 'SubNode',
+			runIndex: 0,
+			data: {
+				data: [{ json: { result: 'from RootNode' } }],
+				inOut: 'output' as const,
+				type: NodeConnectionTypes.AiTool,
+				// No source field intentionally
+				metadata: {
+					executionTime: 100,
+					startTime: Date.parse('2025-02-26T00:00:01.000Z'),
+					subExecution: undefined,
+				},
+			},
+		};
+
+		// Create test AI data array
+		const aiData = [subNodeData];
+
+		// Test for RootNode - should still show SubNode even without source info
+		const rootNodeTree = getTreeNodeData('RootNode', workflow, aiData);
+		expect(rootNodeTree[0].children.length).toBe(1);
+		expect(rootNodeTree[0].children[0].node).toBe('SubNode');
+		expect(rootNodeTree[0].children[0].runIndex).toBe(0);
+	});
+
+	it('should include nodes with empty source array', () => {
+		const workflow = createTestWorkflowObject({
+			nodes: [createTestNode({ name: 'RootNode' }), createTestNode({ name: 'SubNode' })],
+			connections: {
+				SubNode: {
+					ai_tool: [[{ node: 'RootNode', type: NodeConnectionTypes.AiTool, index: 0 }]],
+				},
+			},
+		});
+
+		// Create test AI data with a node that has empty source array
+		const subNodeData = {
+			node: 'SubNode',
+			runIndex: 0,
+			data: {
+				data: [{ json: { result: 'from RootNode' } }],
+				inOut: 'output' as const,
+				type: NodeConnectionTypes.AiTool,
+				source: [], // Empty array
+				metadata: {
+					executionTime: 100,
+					startTime: Date.parse('2025-02-26T00:00:01.000Z'),
+					subExecution: undefined,
+				},
+			},
+		};
+
+		// Create test AI data array
+		const aiData = [subNodeData];
+
+		// Test for RootNode - should still show SubNode even with empty source array
+		const rootNodeTree = getTreeNodeData('RootNode', workflow, aiData);
+		expect(rootNodeTree[0].children.length).toBe(1);
+		expect(rootNodeTree[0].children[0].node).toBe('SubNode');
+		expect(rootNodeTree[0].children[0].runIndex).toBe(0);
+	});
+
+	it('should filter deeper nested nodes based on source', () => {
```

Comment:
Does this work because we filter runs on deeper (2nd+ level) nodes based on the 1st level sub node's run index? Thinking of the case where they don't match like below
<img width="652" alt="Screenshot 2025-05-02 at 10 29 40" src="https://github.com/user-attachments/assets/aed6a3fa-e68b-4f32-a376-ec965d5778cd" />